### PR TITLE
✨ Add environment locking

### LINKED ISSUE #40: Add Conda Environment Locking
To improve reproducibility and stability, repository needs environment locking in the same way as the InnerEye-DeepLearning repo.

### PR DESCRIPTION
**Pull request by peterhessey**
Closes #40 

Adds environment locking to secondary dependencies to prevent unexpected behaviour.

### CODE REVIEW

**Comment by peterhessey** [comment]
Forgotten to add windows instructions and env file - marking as draft until that is complete

File: README.md [hunk 1]
```diff
@@ -172,7 +186,7 @@ During inference the image data zip file is copied to the IMAGE_DATA_FOLDER in t
 
 This project welcomes contributions and suggestions.  Most contributions require you to agree to a
 Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us
-the rights to use your contribution. For details, visit https://cla.opensource.microsoft.com.
+the rights to use your contribution. For details, visit the [the Microsoft CLA site](https://cla.opensource.microsoft.com).
```
Comment (ant0nsc) [inline README.md:189]:
```suggestion
the rights to use your contribution. For details, visit the [Microsoft CLA site](https://cla.opensource.microsoft.com).
```

**Review by ant0nsc — APPROVED** [review]

**Review by ant0nsc — APPROVED** [review]

File: create_and_lock_environment.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        unset python_arr[-1]
+        echo "${python_arr[0]}"="${python_arr[1]}"
+    elif [[ ! $line == "#"* ]]; then
+        echo "${line}"
+    fi
+done < environment.yml > environment.yml.tmp
+echo "# WARNING - DO NOT EDIT THIS FILE MANUALLY" > environment.yml
+echo "# Please refer to the environment documentation for instructions on how to create a new version of this file: https://github.com/microsoft/InnerEye-DeepLearning/blob/main/docs/environment.md" >> environment.yml
+cat environment.yml.tmp >> environment.yml
+rm environment.yml.tmp
```
Comment (fepegar) [inline create_and_lock_environment.sh:39]: Please make sure there's a newline at the end of the file, otherwise pre-commit will complain. You can set this automatically in VS Code.

File: README.md [hunk 3]
```diff
@@ -13,20 +12,32 @@ Download a Conda or Miniconda [installer for your platform](https://docs.conda.i
...
 ### Creating a Conda environment
-Note that in order to create the Conda environment you will need to have build tools installed on your machine. If you are running Windows, they should be already installed with Conda distribution.   
 
-You can install build tools on Ubuntu (and Debian-based distributions) by running  
-`sudo apt-get install build-essential`  
-If you are running CentOS/RHEL distributions, you can install the build tools by running  
+Note that in order to create the Conda environment you will need to have build tools installed on your machine. If you are running Windows, they should be already installed with Conda distribution.
+
+You can install build tools on Ubuntu (and Debian-based distributions) by running
+`sudo apt-get install build-essential`
```
Comment (fepegar) [inline README.md:19]: Period missing.

File: README.md [hunk 4]
```diff
@@ -13,20 +12,32 @@ Download a Conda or Miniconda [installer for your platform](https://docs.conda.i
...
+
 * `conda env create --file environment.yml`
 * `conda activate inference`
 
+#### Windows Users
+
+Start the `conda` prompt for your platform. In that prompt, navigate to your repository root and run
+
+* `conda env create --file environment_win.yml`
+* `conda activate inference`
```
Comment (fepegar) [inline README.md:35]: Better to enclose this in a `console` or `shell` code block, I think.

File: README.md [hunk 5]
```diff
@@ -56,13 +67,13 @@ The app can be tested locally using [`curl`](https://curl.se/).
...
 
-```
+```console
 curl -i -H "API_AUTH_SECRET: <val of CUSTOMCONNSTR_API_AUTH_SECRET>" http://localhost:5000/v1/ping
 ```
 
 This should produce an output similar to:
 
-```
+```console
```
Comment (fepegar) [inline README.md:76]: As this are not console commands, maybe you can just use `text`.

Comment (fepegar) [reply README.md:76]: Also applies to similar lines below.

File: README.md [hunk 6]
```diff
@@ -142,8 +153,9 @@ Date: Wed, 18 Aug 2021 14:01:27 GMT
 and download the inference result as a zipped DICOM-RT file to `HN_rt.zip`.
 
 ### Running flask app in Azure
+
 * Install Azure CLI: `curl -sL https://aka.ms/InstallAzureCLIDeb | sudo bash`
-* Login: `az login --use-device-code`  
+* Login: `az login --use-device-code`
```
Comment (fepegar) [inline README.md:158]: If this is a sequence of instructions, a numbered list might be better. Note you can use "1" for all the points.

File: README.md [hunk 7]
```diff
@@ -158,6 +170,17 @@ If you would like to reproduce the automatic deployment of the service for testi
...
 During inference the image data zip file is copied to the IMAGE_DATA_FOLDER in the AzureML workspace's DATASTORE_NAME datastore. At the end of inference the copied image data zip file is overwritten with a simple line of text. At present we cannot delete these. If you would like these overwritten files removed from your datastore you can [add a policy](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-lifecycle-management-concepts?tabs=azure-portal) to delete items from the datastore after a period of time. We recommend 7 days.
 
+## Changing Primary Dependencies
+
+1. Make your desired changes in `primary_deps.yml`. Make sure your package name and version are correct.
+2. To create a new environment and a valid `environment.yml`, run the following command:
+
+    ```shell
+    bash -i create_and_lock_environment.sh
+    ```
```
Comment (fepegar) [inline README.md:180]: No need to indent this.

File: README.md [hunk 8]
```diff
@@ -186,8 +209,8 @@ contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additio
 
 This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
 
-# Resources:
+## Resources
 
-- [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/)
-- [Microsoft Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/)
-- Contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with questions or concerns
+* [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/)
```
Comment (fepegar) [inline README.md:214]: Out of curiosity, why was this changed? Did the linter complain?

File: primary_deps.yml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+      - flask-injector==0.13.0
+      - Flask==2.1.1
+      - hi-ml-azure==0.2.1
+      - memory-tempfile==2.2.3
+      - mypy==0.790
+      - pydicom==2.1.2
+      - pytest-cov==2.10.1
+      - pytest==6.0.1
+      - SimpleITK==2.0.2
+      - Werkzeug==2.1.1
```
Comment (fepegar) [inline primary_deps.yml:22]: Please add a new line here.

**Review by fepegar — CHANGES_REQUESTED** [review]

**Comment by fepegar** [comment]
Requested minor changes, please re-request the review and I'll approve.

File: create_and_lock_environment.sh [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    echo "Starting AML environment locking..."
+fi
+
+# get environment name from primary dependencies YAML file
+name_line="$(cat primary_deps.yml | grep 'name:')"
+IFS=':' read -ra name_arr <<< "$name_line"
+env_name="${name_arr[1]}"
+
+# clear old conda envs, create new one
+export CONDA_ALWAYS_YES="true"
```
[the remark this answers is not in the document]
Comment (peterhessey) [reply create_and_lock_environment.sh:17]: I think we concluded in the other PR that it is necessary, as otherwise conda prompts for approval when running the script which is somewhat annoying,

[the remark this answers is not in the document]
Comment (fepegar) [reply create_and_lock_environment.sh:17]: Oh I might gave missed that discussion. Doesn't it work if you simply set the variable with exporting and unsetting? That's my suggestion.

File: README.md (hunk 8 above)
Comment (fepegar) [reply README.md:218]: Oh sorry, I meant `-` vs `*` for Markdown bullet points.

**Review by fepegar — APPROVED** [review]
